File: src/sap.nim
import std/[deques, strformat, tables], form, jsony

proc addSap(str: var string, r: Ref) =
  str.add(&"[{r.kind.ord}")
  case r.kind:
  of Tag:
    str.add(&",\"{r.tag}\",")
    str.addSap(r.tagged)
  of Undef, Null: discard
  of Bool:
    if r.b:
      str.add(",true")
    else:
      str.add(",false")
  of Sym:
    str.add(&",{r.sym.toJSON()}")
  of Str:
    str.add(&",{r.str.toJSON()}")
  of Bin:
    str.add(&",{r.bin.toJSON()}")
  of Vec:
    for e in r.vec.items:
      str.add(',')
      str.addSap(e)
  of Map:
    for k, v in r.map.pairs:
      str.add(',')
      str.addSap(k)
      str.add(',')
      str.addSap(v)
  of U64:
    str.add(&",\"{r.u64}\"")
  of I64:
    str.add(&",\"{r.i64}\"")
  of F64:
    str.add(&",\"{r.f64}\"")
  of U32:
    str.add(&",\"{r.u32}\"")
  of I32:
    str.add(&",\"{r.i32}\"")
  of F32:
    str.add(&",\"{r.f32}\"")
  of U16:
    str.add(&",\"{r.u16}\"")
  of I16:
    str.add(&",\"{r.i16}\"")
  of F16:
    str.add(&",\"{r.f16}\"")
  of U8:
    str.add(&",\"{r.u8}\"")
  of I8:
    str.add(&",\"{r.i8}\"")
  str.add("]")


proc sap*(r: Ref): string =
  result.addSap(r)

File: src/sap.ts
enum Type {
  Tag = 0,
  Undef = 1,
  Null = 2,
  Bool = 3,
  Sym = 8,
  Str = 9,
  Bin = 10,
  Vec = 11,
  Map = 12,
  U64 = 16,
  U32 = 17,
  U16 = 18,
  U8 = 19,
  I64 = 32,
  I32 = 33,
  I16 = 34,
  I8 = 35,
  F64 = 48,
  F32 = 49,
  F16 = 50,
}

type Tag = [Type.Tag, Sap];
type Undef = [Type.Undef];
type Null = [Type.Null];
type Bool = [Type.Bool, boolean];
type Sym = [Type.Sym, string];
type Str = [Type.Str, string];
type Bin = [Type.Bin, string];
type Vec = [Type.Vec, ...Sap[]];
type SMap = [Type.Map, ...Sap[]];
type U64 = [Type.U64, string];
type U32 = [Type.U32, string];
type U16 = [Type.U16, string];
type U8 = [Type.U8, string];
type I64 = [Type.I64, string];
type I32 = [Type.I32, string];
type I16 = [Type.I16, string];
type I8 = [Type.I8, string];
type F64 = [Type.F64, string];
type F32 = [Type.F32, string];
type F16 = [Type.F16, string];


type Sap =
| Tag
| Undef | Null
| Bool
| Sym | Str | Bin
| Vec | SMap
| U64 | U32 | U16 | U8
| I64 | I32 | I16 | I8
| F64 | F32 | F16;
